admin/jobs/river/biller_event_handlers.go

```diff
@@ -0,0 +1,286 @@
+		if errors.Is(err, database.ErrNotFound) {
+			// org got deleted, ignore
+			return nil
+		}
+		return fmt.Errorf("failed to find organization: %w", err)
+	}
+
+	if time.Now().UTC().Before(job.Args.GracePeriodEndDate.AddDate(0, 0, 1)) {
+		return fmt.Errorf("grace period date %s not finished yet for org %q", job.Args.GracePeriodEndDate, org.Name) // will be retried later
+	}
+
+	// check if the org has still invoice failed billing error
+	be, err := w.admin.DB.FindBillingErrorByType(ctx, org.ID, database.BillingErrorTypeInvoicePaymentFailed)
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			return fmt.Errorf("failed to find billing errors: %w", err)
+		}
+	}
+
+	if be != nil {
+		// just to be very sure, check if the invoice is still unpaid directly from the biller
+		invoice, err := w.admin.Biller.GetInvoice(ctx, job.Args.InvoiceID)
+		if err != nil {
+			return fmt.Errorf("failed to get invoice %q: %w", job.Args.InvoiceID, err)
+		}
+
+		if w.admin.Biller.IsInvoicePaid(ctx, invoice) || !w.admin.Biller.IsInvoiceValid(ctx, invoice) {
+			w.admin.Logger.Warn("Invoice was already paid or invalid but billing error was not cleared", zap.String("org_id", org.ID), zap.String("org_name", org.Name), zap.String("billing_customer_id", org.BillingCustomerID), zap.String("invoice_id", job.Args.InvoiceID), zap.String("invoice_status", invoice.Status))
+
+			// clearing the billing error for this invoice
+			failedInvoices := be.Metadata.(*database.BillingErrorMetadataInvoicePaymentFailed).Invoices
+			delete(failedInvoices, job.Args.InvoiceID)
+
+			// if no more failed invoices, delete the billing error
+			if len(failedInvoices) == 0 {
+				err = w.admin.DB.DeleteBillingError(ctx, be.ID)
+				if err != nil {
+					return fmt.Errorf("failed to delete billing error: %w", err)
+				}
+			} else {
+				// update the metadata
+				_, err = w.admin.DB.UpsertBillingError(ctx, &database.UpsertBillingErrorOptions{
+					OrgID:     org.ID,
+					Type:      database.BillingErrorTypeInvoicePaymentFailed,
+					Metadata:  &database.BillingErrorMetadataInvoicePaymentFailed{Invoices: failedInvoices},
+					EventTime: be.EventTime,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to update billing error: %w", err)
+				}
+			}
+			return nil
+		}
+
+		// hibernate projects
+		limit := 10
+		afterProjectName := ""
+		for {
+			projs, err := w.admin.DB.FindProjectsForOrganization(ctx, org.ID, afterProjectName, limit)
+			if err != nil {
+				return err
+			}
+
+			for _, proj := range projs {
+				_, err = w.admin.HibernateProject(ctx, proj)
+				if err != nil {
+					return fmt.Errorf("failed to hibernate project %q: %w", proj.Name, err)
+				}
+				afterProjectName = proj.Name
+			}
+
+			if len(projs) < limit {
+				break
+			}
+		}
+		w.admin.Logger.Info("Projects hibernated due to no valid payment method", zap.String("org_id", org.ID), zap.String("org_name", org.Name), zap.String("billing_customer_id", org.BillingCustomerID))
```

Comment:
Is there any way we can namespace all these billing-related logs? Either with a `Named` logger or with a `billing: ...` prefix in the messages. It might be nice to setup a Slack channel of all billing-related logs.

Reply:
nice idea, added a named logger